panoramapublic/src/org/labkey/panoramapublic/pipeline/AssignSubmitterPermissionJob.java

```diff
@@ -0,0 +1,135 @@
+        if (_project != null)
+        {
+            assignRole();
+            setStatus(PipelineJob.TaskStatus.complete);
+        }
+        else
+        {
+            getLogger().error("Input project was null. Exiting...");
+        }
+    }
+
+    private void assignRole()
+    {
+        var containers = ContainerManager.getAllChildren(_project, getUser());
+        getLogger().info("Total number of folders: " + containers.size());
+
+        int done = 0;
+        int updated = 0;
+        int notExptFolders = 0;
+        int total = containers.size();
+
+        try (DbScope.Transaction transaction = PanoramaPublicManager.getSchema().getScope().ensureTransaction())
+        {
+            for (Container container : containers)
+            {
+                ExperimentAnnotations expAnnotations = ExperimentAnnotationsManager.getExperimentInContainer(container);
+                if (expAnnotations != null && expAnnotations.isJournalCopy())
+                {
+                    boolean submitterUpdated = assignRole(expAnnotations.getSubmitterUser(), "Submitter", container, _dryRun, getLogger());
+                    boolean labHeadUpdated = assignRole(expAnnotations.getLabHeadUser(), "Lab Head", container, _dryRun, getLogger());
+                    if (submitterUpdated || labHeadUpdated)
+                    {
+                        updated++;
+                    }
+                }
+                else
+                {
+                    getLogger().info(String.format("'%s' - no valid experiment", container.getPath()));
+                    notExptFolders++;
+                }
+
+                done++;
+                if (done % 100 == 0)
+                {
+                    getLogger().info(done + "/" + total + " done.");
+                }
+            }
+            transaction.commit();
+        }
+
+        getLogger().info("All folders: " + total + "; Folders with valid experiments: " + (total - notExptFolders));
+        getLogger().info("Assigned PanoramaPublicSubmitterRole in " + updated + " folders.");
+    }
+
+    private boolean assignRole(User user, String userType, Container container, boolean dryRun, Logger logger)
+    {
+        if (user == null) return false;
+
+        if (!container.hasPermission(user, PanoramaPublicSubmitterPermission.class))
+        {
+            if (!dryRun)
+            {
+                logger.info(String.format("'%s', %s: %s - %s", container.getPath(), userType, user.getEmail(), "assigning"));
+                MutableSecurityPolicy newPolicy = new MutableSecurityPolicy(container, container.getPolicy());
+                newPolicy.addRoleAssignment(user, PanoramaPublicSubmitterRole.class, false);
+                SecurityPolicyManager.savePolicy(newPolicy);
+                return true;
+            }
+            else
+            {
+                logger.info(String.format("'%s', %s: %s - %s", container.getPath(), userType, user.getEmail(), "would assign (dry run)"));
```

Comment:
From my reading of the code, it should be iterating through the containers depth-first. That's good, it means that the role assignment may be added at the project level and then cascade down any children that inherit permissions.

However, I think the dry run will log that it's adding it to all children even though it won't need to. If you don't care, that's fine, but you could track the resource ID of the policy used for a container and not log (or log differently) when you revisit one.

Reply:
It should be OK since the role will only be assigned in containers that have a TargetedMSExperiment.  All other containers will be logged as not having a valid experiment.